src/scipp/coords.py

```diff
@@ -2,247 +2,480 @@
+class Graph:
+    def __init__(self, graph):
+        self._graph = _convert_to_rule_graph(graph)
+
+    def subgraph(self, data: DataArray, targets: Set[str]) -> Dict[str, _Rule]:
+        subgraph = {}
+        pending = list(targets)
+        while pending:
+            out_name = pending.pop()
+            if out_name in subgraph:
+                continue
+            rule = self._rule_for(out_name, data)
+            subgraph[out_name] = rule
+            pending.extend(rule.dependencies)
+        return subgraph
+
+    def _rule_for(self, out_name: str, data: DataArray) -> _Rule:
+        if _is_meta_data(out_name, data):
+            return _FetchRule((out_name, ), data.meta,
+                              data.bins.meta if data.bins else None)
         try:
-            from graphviz import Digraph
-        except ImportError:
-            warnings.warn("Failed to import `graphviz`, please install `graphviz` if "
-                          "using `pip`, or `python-graphviz` if using `conda`.")
-        dot = Digraph(strict=True)
+            return self._graph[out_name]
+        except KeyError:
+            raise NotFoundError(
+                f"Coordinate '{out_name}' does not exist in the input data "
+                "and no rule has been provided to compute it.") from None
+
+    def show(self, size=None, simplified=False):
+        dot = _make_digraph(strict=True)
         dot.attr('node', shape='box', height='0.1')
         dot.attr(size=size)
-        for output, producer in self._graph.items():
-            if isinstance(producer, str):  # rename
-                dot.edge(producer, output)
-            else:
+        for output, rule in self._graph.items():
+            if isinstance(rule, _RenameRule):
+                dot.edge(rule.dependencies[0], output, style='dashed')
+            elif isinstance(rule, _ComputeRule):
                 if not simplified:
-                    name = f'{producer.__name__}(...)'
+                    name = f'{rule.func_name}(...)'
                     dot.node(name, shape='ellipse', style='filled', color='lightgrey')
                     dot.edge(name, output)
                 else:
                     name = output
-                argnames = _argnames(producer)
-                for arg in argnames:
+                for arg in rule.dependencies:
                     dot.edge(arg, name)
         return dot
 
 
-def _consume_coord(obj, name):
-    if name in obj.coords:
-        obj.attrs[name] = obj.coords.pop(name)
-    if obj.bins is not None:
-        if name in obj.bins.coords:
-            obj.bins.attrs[name] = obj.bins.coords.pop(name)
-        return obj.attrs.get(name, None), obj.bins.attrs.get(name, None)
-    return obj.attrs[name], None
+def _log_plan(rules: List[_Rule], targets: Set[str],
+              dim_name_changes: Mapping[str, str], coords: _CoordTable) -> None:
+    inputs = set(_rule_output_names(rules, _FetchRule))
+    bi_products = {
```

Comment:
bi or by?

Reply:
That is why the spellchecker kept complaining! :smiley: